internal/rbd/encryption.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2021 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rbd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/ceph/ceph-csi/internal/util"
+
+	librbd "github.com/ceph/go-ceph/rbd"
+)
+
+// rbdEncryptionState describes the status of the process where the image is
+// with respect to being encrypted.
+type rbdEncryptionState string
+
+const (
+	// rbdImageEncryptionUnknown means the image is not encrypted, or the
+	// metadata of the image can not be fetched.
+	rbdImageEncryptionUnknown = rbdEncryptionState("")
+	// rbdImageEncrypted is set in the image metadata after the image has
+	// been formatted with cryptsetup. Future usage of the image should
+	// unlock the image before mounting.
+	rbdImageEncrypted = rbdEncryptionState("encrypted")
+	// rbdImageEncryptionPrepared gets set in the image metadata once the
+	// passphrase for the image has been generated and stored in the KMS.
+	// When using the image for the first time, it needs to be encrypted
+	// with cryptsetup before updating the state to `rbdImageEncrypted`.
+	rbdImageEncryptionPrepared = rbdEncryptionState("encryptionPrepared")
+
+	// rbdImageRequiresEncryption has been deprecated, it is used only for
+	// volumes that have been created with an old provisioner, were never
+	// attached/mounted and now get staged by a new node-plugin
+	// TODO: remove this backwards compatibility support
+	rbdImageRequiresEncryption = rbdEncryptionState("requiresEncryption")
+
+	// image metadata key for encryption
+	encryptionMetaKey = ".rbd.csi.ceph.com/encrypted"
```

Comment:
@nixpanic may be use this constant here too https://github.com/ceph/ceph-csi/blob/d534ee9ce8b81b731fab7d5e9c12816ad3b8e511/e2e/rbd_helper.go#L246

Reply:
We do not exposed any internals to the e2e tests at the moment. It would be an improvement, I guess,. That also suggests we should open and document the API (or at least metadata keys/values, JSON structures, ...). We can do that, but I would prefer to handle it through an other issue (please file one).

Reply:
Sure. it can be another PR or issue. Thanks!